### PR TITLE
Add onBlur event

### PR DESCRIPTION
**Pull request by megantfanning**
## Reasons for making this change

Fixes issue #364. Adds *onBlur* functionality to react json schema form allowing user to trigger validation or saving on field focus changes.

Note: There is an additional pull request #431, which provides similar functionality but additionally returns the id of the field which triggered the `onBlur` event. We have figured we would defer to the maintainers preferences on how to implement the `onBlur` event and submit in addition.

### Checklist

* [X] **I'm updating documentation**
  - [X] I've checked the rendering of the Markdown text I've added
  - [X] If I'm adding a new section, I've updated the Table of Content
* [X] **I'm adding or updating code**
  - [X] I've added and/or updated tests
  - [X] I've updated docs if needed
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

File: dist/react-jsonschema-form.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
...
...bol"==("undefined"==typeof e?"undefined":U(e))||D(e)&&le.call(e)==J}function W(e){return null==e?"":j(e)}function $(e){return ge(e)?r(e,F):q(e)?[e]:x(ve(e))}var U="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e},B="Expected a function",L="__lodash_hash_undefined__",z=1/0,Z="[object Function]",V="[object GeneratorFunction]",J="[object Symbol]",H=/^\./,G=/[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g,K=/[\\^$.*+?()[\]{}|]/g,X=/\\(\\)?/g,Y=/^\[object .+?Constructor\]$/,Q="object"==("undefined"==typeof t?"undefined":U(t))&&t&&t.Object===Object&&t,ee="object"==("undefined"==typeof self?"undefined":U(self))&&self&&self.Object===Object&&self,te=Q||ee||Function("return this")(),re=Array.prototype,ne=Function.prototype,oe=Object.prototype,ae=te["__core-js_shared__"],ie=function(){var e=/[^.]+$/.exec(ae&&ae.keys&&ae.keys.IE_PROTO||"");return e?"Symbol(src)_1."+e:""}(),ue=ne.toString,se=oe.hasOwnProperty,le=oe.toString,ce=RegExp("^"+ue.call(se).replace(K,"\\$&").replace(/hasOwnProperty|(func
```
Comment (n1k0) [inline dist/react-jsonschema-form.js]: Dist files should not be versioned in this repository. They're generated locally before being published to npm.

File: lib/components/ErrorList.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      errors.map(function (error, i) {
+        return _react2.default.createElement(
+          "li",
+          { key: i, className: "list-group-item text-danger" },
+          error.stack
+        );
+      })
+    )
+  );
+}
```
Comment (n1k0) [inline lib/components/ErrorList.js]: Same for lib files, they're generated locally then published to npm, and should therefore not be versioned in this repository.

File: playground/app.js [hunk 3]
```diff
@@ -328,7 +328,8 @@ class App extends Component {
             <div className="col-sm-2">
               <Form schema={liveValidateSchema}
                     formData={liveValidate}
-                    onChange={this.setLiveValidate}><div/></Form>
+                    onChange={this.setLiveValidate}
+                    onBlur={(event)=> console.log("onBlur:", event)}><div/></Form>
```
Comment (n1k0) [inline playground/app.js]: We could probably remove this.

File: playground/app.js [hunk 4]
```diff
@@ -357,6 +358,7 @@ class App extends Component {
               uiSchema={uiSchema}
               formData={formData}
               onChange={this.onFormDataChange}
+              onBlur={x => console.log("onBlur", x)}
```
Comment (n1k0) [inline playground/app.js]: Ditto.

File: src/components/Form.js [hunk 5]
```diff
@@ -146,15 +145,15 @@ export default class Form extends Component {
...
-        id={id}
-        name={name}
-        method={method}
-        target={target}
-        action={action}
-        autoComplete={autocomplete}
-        encType={enctype}
-        acceptCharset={acceptcharset}
-        onSubmit={this.onSubmit}>
+            id={id}
```
Comment (n1k0) [inline src/components/Form.js]: Please keep indentation the way it is.

File: src/components/widgets/AltDateWidget.js [hunk 6]
```diff
@@ -16,7 +16,7 @@ function readyForChange(state) {
 }
 
 function DateElement(props) {
-  const {type, range, value, select, rootId, disabled, readonly, autofocus, registry} = props;
+  const {type, range, value, select, rootId, disabled, readonly, autofocus, registry,onBlur} = props;
```
Comment (n1k0) [inline src/components/widgets/AltDateWidget.js]: Nit: missing space after comma.

File: src/components/widgets/AltDateWidget.js [hunk 7]
```diff
@@ -63,6 +64,12 @@ class AltDateWidget extends Component {
     });
   };
 
+  onBlur = () => {
+    if (this.props.onBlur){
+      this.props.onBlur(toDateString(this.state, this.props.time));
+    }
+  };
```
Comment (n1k0) [inline src/components/widgets/AltDateWidget.js]: Don't we want to trigger the event when data are consistent, the same way it's done in `onChange`?

File: .gitignore [hunk 8]
```diff
@@ -1,6 +1,4 @@
 npm-debug.log
 node_modules
 build
-dist
-lib
-
+.idea
```
Comment (n1k0) [inline .gitignore]: This should be added by IntelliJ users in their own global `~/.gitignore` file. Also please readd ignore rules for `dist` and `lib` (see contextual comments below).

File: src/components/widgets/BaseInput.js [hunk 9]
```diff
@@ -22,7 +23,12 @@ function BaseInput(props) {
       readOnly={readonly}
       autoFocus={autofocus}
       value={typeof value === "undefined" ? "" : value}
-      onChange={(event) => onChange(event.target.value)}/>
+      onChange={(event) => onChange(event.target.value)}
+      onBlur={(event) => {
+        if (onBlur) {
+          return onBlur(event.target.id);
```
Comment (n1k0) [inline src/components/widgets/BaseInput.js]: Why don't we pass the value to the event handler here?

**Comment by jbalboni** [comment]
It's ultimately not up to me, obviously, but there aren't really substantial functional differences between my other PR for this and this one. Any changes you wanted could have been been brought up comments and I would have been happy to follow whatever path @n1k0 thought was best. It would have made the review process less intensive, since the diff for my PR is less noisy.

**Comment by n1k0** [comment]
I agree, I'd love seeing efforts federated for this feature :+1:

**Comment by n1k0** [comment]
Closing in favor of #431. Please reopen if you have strong arguments against this decision.

**Pull request closed without merging**